libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,827 @@
+      }
+    }
+  }
+
+  auto find_gap(const data& curr, const data* prev) const
+    -> std::optional<data> {
+    if (not prev) {
+      return std::nullopt;
+    }
+    return match(
+      std::tie(curr, *prev),
+      [&](const duration& c, const duration& p) -> std::optional<data> {
+        if (c - p > args_.res->inner) {
+          return p + args_.res->inner;
+        }
+        return std::nullopt;
+      },
+      [&](const time& c, const time& p) -> std::optional<data> {
+        if (c - p > args_.res->inner) {
+          return p + args_.res->inner;
+        }
+        return std::nullopt;
+      },
+      [](const auto&, const auto&) -> std::optional<data> {
+        TENZIR_UNREACHABLE();
+      });
+  }
+
+  auto make_attributes(
+    const std::string& xpath, std::deque<std::string>& ynums,
+    std::unordered_map<std::string, bool>& ynames,
+    const detail::stable_map<std::string_view, std::string>& limits) const
+    -> std::vector<type::attribute_view> {
+    auto attrs = std::vector<type::attribute_view>{
+      {"chart", to_string(args_.ty)},
+      {"position", args_.position.inner},
+      {"x_axis_type", args_.x_log ? "log" : "linear"},
+      {"y_axis_type", args_.y_log ? "log" : "linear"},
+      {"x", xpath},
+    };
+    for (const auto& [name, value] : limits) {
+      if (not value.empty()) {
+        attrs.emplace_back(name, value);
+      }
+    }
+    for (auto i = ynums.size(); i < args_.y.size() or i < ynames.size(); ++i) {
+      ynums.emplace_back(fmt::format("y{}", i));
+    }
+    auto names = std::views::filter(ynames, [](auto&& x) {
+      return x.second;
+    });
+    for (const auto& [num, field] : detail::zip(ynums, names)) {
+      attrs.emplace_back(num, field.first);
+    }
+    return attrs;
+  }
+
+  auto validate_x(const type& ty, diagnostic_handler& dh) const -> bool {
+    auto valid = ty.kind()
+                   .is_any<int64_type, uint64_type, double_type, duration_type,
+                           time_type>();
+    if (args_.ty == chart_type::bar or args_.ty == chart_type::pie) {
+      valid |= ty.kind().is_any<null_type, ip_type, subnet_type, string_type>();
+    }
+    if (not valid) {
+      diagnostic::warning("x-axis cannot have type `{}`", ty.kind())
+        .note("skipping invalid events")
+        .primary(args_.x)
+        .emit(dh);
+      return false;
+    }
+    if (args_.res and not ty.kind().is_any<time_type, duration_type>()) {
+      diagnostic::warning("cannot group type `{}` with resolution", ty.kind())
+        .note("skipping invalid events")
+        .primary(args_.x)
+        .primary(args_.res->source)
+        .emit(dh);
+      return false;
+    }
+    return true;
+  }
+
+  auto validate_y(const data& d, tenzir::location loc,
+                  diagnostic_handler& dh) const -> bool {
+    if (args_.ty == chart_type::line and is<caf::none_t>(d)) {
+      return true;
+    }
+    const auto ty = type::infer(d);
+    if (not ty) {
+      return false;
```

Comment:
Does this lead to a diagnostic somewhere or what happens in this case?

Reply:
It skips charting that `y` series while still keeping the data. I think it's unlikely for it to ever trigger but do you think this should be a warning like below?

Reply:
Every failure path should have either a diagnostic or an assertion. Can you try to eliminate the failure path by not using `type::infer`, but instead directly looking at what `data` holds?

Reply:
That does not give me a representation of the type for emitting in the diagnostic unless I do nearly the same thing as `type::infer`.